### PR TITLE
Feature/sparsestore open modes

### PR DESCRIPTION
**Pull request by karimyoussef91**
Supporting create mode, open mode, with and without read_ony.

### CODE REVIEW

File: tests/umap-sparsestore/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,35 @@
+
+#############################################################################
+# Copyright 2017-2019 Lawrence Livermore National Security, LLC and other
```
Comment (mayagokhale) [inline tests/umap-sparsestore/CMakeLists.txt]: update to 2021 copyright

File: src/umap/store/SparseStore.h [hunk 2]
```diff
@@ -1,4 +1,3 @@
-  
 //////////////////////////////////////////////////////////////////////////////
 // Copyright 2017-2020 Lawrence Livermore National Security, LLC and other
```
Comment (mayagokhale) [inline src/umap/store/SparseStore.h]: update to 2021 copyright

**Review by mayagokhale — DISMISSED** [review]
please update the copyrights to 2021

File: tests/umap-sparsestore/CMakeLists.txt (hunk 1 above)
Comment (karimyoussef91) [reply tests/umap-sparsestore/CMakeLists.txt]: I've updated the copyrights and pushed the changes.
Thanks!

File: src/umap/store/SparseStore.h (hunk 2 above)
Comment (karimyoussef91) [reply src/umap/store/SparseStore.h]: I've updated the copyrights and pushed the changes.
Thanks!

File: src/umap/store/SparseStore.cpp [hunk 3]
```diff
@@ -40,41 +44,61 @@ namespace Umap {
...
+        std::ifstream metadata(metadata_file_path.c_str());
         if (!metadata.is_open()){
           UMAP_ERROR("Failed to open metadata file" << " - " << strerror(errno));
         }
         else{
-          metadata << file_size;
+          metadata >> file_size;
+          metadata >> current_capacity;
+          numreads = numwrites = 0;
+          num_files = num_files = (uint64_t) ceil( current_capacity*1.0 / file_size );
```
Comment (asarkar-parsys) [inline src/umap/store/SparseStore.cpp]: num_files = num_files?

File: docs/sphinx/sparse_store.rst [hunk 4]
```diff
@@ -4,45 +4,22 @@
...
-      size_t file_size){
-
-     void * region = NULL;
      
-     // Instantiating a SparseStore objct, the "file_size" parameter specifies the granularity of each file. 
-     // An application that desires to create N files can calculate the file size using totalbytes / num_files
-     // Note that the file size is rounded to be a multiple of the page size, 
-     // which results in a number of files that close but not exactly equal to N
-     Umap::SparseStore* sparse_store;
+     Umap::SparseStore sparse_store;
```
Comment (asarkar-parsys) [inline docs/sphinx/sparse_store.rst]: Though in documentation. * maybe missing here?

File: src/umap/store/SparseStore.cpp [hunk 5]
```diff
@@ -114,32 +134,59 @@ namespace Umap {
...
-                if( (fallocate_status = posix_fallocate(fd,0,file_size) ) == 0){
-                  file_descriptors[fd_index].id = fd;
-                }
-                else{
-                  UMAP_ERROR("SparseStore: fallocate() failed for file with id: " << fd_index << " - " << strerror(errno));
+              int flags = (read_only ? O_RDONLY :  O_RDWR ) | O_CREAT | O_DIRECT | O_LARGEFILE;
+              int fd = open(filename.c_str(), flags, S_IRUSR | S_IWUSR);
+              if (fd == -1){
+                // retry without O_DIRECT
+                flags = (read_only ? O_RDONLY :  O_RDWR ) | O_CREAT | O_LARGEFILE;
```
Comment (asarkar-parsys) [inline src/umap/store/SparseStore.cpp:175]: Can you explain when this retry is needed?

Comment (karimyoussef91) [reply src/umap/store/SparseStore.cpp:175]: Some file systems do not support "O_DIRECT" (e.g. TMPFS). So it first tries open() with O_DIRECT, and if fails, it retries without O_DIRECT to eliminate this possibility. If still fails, it throws an error.

File: docs/sphinx/sparse_store.rst (hunk 4 above)
Comment (karimyoussef91) [reply docs/sphinx/sparse_store.rst]: Fixed and pushed

File: src/umap/store/SparseStore.cpp (hunk 3 above)
Comment (karimyoussef91) [reply src/umap/store/SparseStore.cpp]: Fixed and pushed

File: src/umap/store/SparseStore.cpp (hunk 5 above)
Comment (asarkar-parsys) [reply src/umap/store/SparseStore.cpp:175]: Consider adding a comment here? "Handling FS that do not support O_DIRECT, e.g., TMPFS"

**Review by asarkar-parsys — DISMISSED** [review]
Can you address the final comment, squash all your commits into one before merging?

**Comment by karimyoussef91** [comment]
Addressed comments and squashed all commits into a single commit.
Thanks!

**Review by asarkar-parsys — APPROVED** [review]

**Pull request merged**